### PR TITLE
Add component to notify Org member that they need a Code host connection

### PR DESCRIPTION
**Pull request by rafax**
Implemented a banner to ask an Org member to add a personal Code Host connection ([CLOUD-147](https://sourcegraph.atlassian.net/browse/CLOUD-147)) and used it on Org repositories page.

**Screenshots:**
One host:
Screenshot 2021-12-15 at 16 35 00
Two hosts:
Screenshot 2021-12-15 at 16 34 43

Banner in Search results (shown only when search in org context returned 0 results **and** user is a member of the org):
Screenshot 2021-12-16 at 15 01 50

### CODE REVIEW

**Review by unknwon — APPROVED** [review]
The screenshots are looking good to me 😁

**Comment by rafax** [comment]
> The screenshots are looking good to me 😁

That's thanks to @quinnkeast :) I'll add one more reviewer for React code, as I'm not that experienced with React Hooks.

File: client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                        }{' '}
+                        to start searching across the {orgName} organization's private repositories on Sourcegraph.
+                    </p>
+                    <Link className="btn btn-primary" to={`/users/${user.username}/settings/code-hosts`}>
+                        Connect with {userMissing.length === 1 ? userMissing[0] : 'code hosts'}
+                    </Link>
+                </Container>
+            )}
+        </>
+    )
```
Comment (kopancek) [inline client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx:50]: IMO extracting the logic outside of JSX would improve readability of the code. 
E.g. we only want to render something if the user has missing connections.

We can also simplify the JSX a little by extracting the missing code hosts to a string variable.

Lastly, should we use some existing component for an alert here? E.g. an info or warning message?

```suggestion
 if (userMissing.length > 0) {
    const missingString = userMissing.join(' and ')
    return (
        <Container className="mb-4">
                    <h3>Just one more step...</h3>
                    <p>
                        Connect with {missingString} to start searching across the {orgName} private repositories on Sourcegraph.
                    </p>
                    <Link className="btn btn-primary" to={`/users/${user.username}/settings/code-hosts`}>
                        Connect with {missingString}
                    </Link>
                </Container>
            )
        } else {
            return null
        }
```

File: client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+export const OrgUserNeedsCodeHost: React.FunctionComponent<OrgUserNeedsCodeHost> = ({
+    orgExternalServices,
+    user,
+    orgName,
+}) => {
+    const { externalServices: userExternalServices } = useExternalServices(user.id)
+    const orgKinds = orgExternalServices.map(service => service.kind)
+    const userKinds = new Set((userExternalServices || []).map(service => service.kind))
+    const userMissing = orgKinds.filter(kind => !userKinds.has(kind)).map(kind => defaultExternalServices[kind].title)
```
[the remark this answers is not in the document]
Comment (rafax) [reply client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx:29]: Makes sense, it's shorter and not less readable.

File: client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx (hunk 1 above)
Comment (rafax) [reply client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx:50]: Excalidraw for this https://app.excalidraw.com/s/4Dr1S6qmmY7/6YePyrn2Dmv has a structure with a header, body and link at the bottom, so I used an approach similar to what Quinn did in https://github.com/sourcegraph/sourcegraph/pull/28839/files

File: client/web/src/user/settings/repositories/SettingsRepositoriesPage.tsx [hunk 3]
```diff
@@ -310,6 +313,10 @@ export const SettingsRepositoriesPage: React.FunctionComponent<Props> = ({
             )}
             {!isUserOwner && shouldDisplayContextBanner && owner.name && getSearchContextBanner(owner.name)}
             {isErrorLike(status) && <ErrorAlert error={status} icon={true} />}
+            {!isUserOwner && externalServices && user && owner.name && (
+                <OrgUserNeedsCodeHost user={user} orgExternalServices={externalServices} orgName={owner.name} />
+            )}
```
[the remark this answers is not in the document]
Comment (rafax) [reply client/web/src/user/settings/repositories/SettingsRepositoriesPage.tsx:318]: I have a separate PR coming that will include this in Search results.

File: client/web/src/user/settings/repositories/SettingsRepositoriesPage.tsx [hunk 4]
```diff
@@ -71,6 +73,7 @@ export const SettingsRepositoriesPage: React.FunctionComponent<Props> = ({
     const isUserOwner = owner.type === 'user'
     const fetchRepositories = isUserOwner ? listUserRepositories : listOrgRepositories
     const fetchRepositoriesCount = isUserOwner ? fetchUserRepositoriesCount : fetchOrgRepositoriesCount
+    const user = useObservable(authenticatedUser)
```
[the remark this answers is not in the document]
Comment (rafax) [reply client/web/src/user/settings/repositories/SettingsRepositoriesPage.tsx]: Passed down for OrgSettingsArea, we still need useObservable.

File: client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    return null
+}
+
+export interface SearchUserNeedsCodeHost {
+    orgSearchContext?: string
+    user: UserContext
+}
+
+export const SearchUserNeedsCodeHost: React.FunctionComponent<SearchUserNeedsCodeHost> = ({
```
Comment (kopancek) [inline client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx:58]: It's usually a good practice in React to put each component into a separate file

File: client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    const orgName = orgSearchContext.replace('@', '')
+    // Is current user a member of the org that owns the context?
+    const org = (user.organizations?.nodes || []).find(org => org.name === orgName)
+    if (!org) {
+        return null
+    }
+    return (
+        <>
+            <PotentialOrgUserNeedsCodeHost user={user} org={org} />
+        </>
```
Comment (kopancek) [inline client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx:75]: No need to wrap with `<>` here:
```suggestion
            <PotentialOrgUserNeedsCodeHost user={user} org={org} />
```

File: client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+const PotentialOrgUserNeedsCodeHost: React.FunctionComponent<PotentialOrgUserNeedsCodeHost> = ({ org, user }) => {
+    const { externalServices: orgExternalServices } = useExternalServices(org.id)
+    return (
+        <>
+            <OrgUserNeedsCodeHost
+                user={user}
+                orgDisplayName={org.displayName || org.name}
+                orgExternalServices={orgExternalServices}
+            />
+        </>
```
Comment (kopancek) [inline client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx:93]: I think we can include this in `SearchUserNeedsCodeHost` component. From my point of view, that component is responsible for determining if we need to render `OrgUserNeedsCodeHost`.

And no need to wrap it in `<>` here

File: client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx (hunk 5 above)
Comment (rafax) [reply client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx:58]: I did a quick search on our repo, and it seems the convention is mixed https://sourcegraph.com/search?q=context:global+functioncomponent+repo:github.com/sourcegraph&patternType=literal
Let's figure out what to do about the private component below and I'll try to split.

File: client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx (hunk 7 above)
Comment (rafax) [reply client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx:91]: If I try to include this in SearchUserNeedsCodeHost, I get a:
```
React Hook "useExternalServices" is called conditionally. React Hooks must be called in the exact same order in every component render. Did you accidentally call a React Hook after an early return?
``` 
Removed <>.

Comment (kopancek) [reply client/web/src/user/settings/codeHosts/OrgUserNeedsCodeHost.tsx]: Right, forgot about that shenanigan 🤦

**Review by kopancek — APPROVED** [review]

**Comment by rafax** [comment]
Main dry-run passed here https://github.com/sourcegraph/sourcegraph/pull/29180, merging.

**Pull request merged**